csrc/scheduler/transpose.cpp

```diff
@@ -30,6 +30,30 @@ namespace nvfuser {
 
 namespace {
 
+struct TransposeViewPropagator : public MaxInfoSpanningTree::Propagator {
```

Comment:
Can you please add some comments here? For example, what is `p2c_via_view` supposed to do?

Reply:
put some comments here~~

Reply:
Thanks. Could you add a test to confirm an unsupported fusion is rejected by this check?

Reply:
I do have a test that validate this. I have renamed the test to better reflect it.
`ReshapePermuteTransposeSchedulerRejectByTransposeViewPropagator_CUDA`

Reply:
This part is the only remaining part I'm still thinking about. I was expecting we should use `requiresForwardViewReplay` or something similar. Is this check used as a quick alternative?

Reply:
Interesting. I wasn't aware that we have a similar check.

However, I do believe their behavior to differ.
Looking at `bool requiresForwardViewReplay(Fusion* fusion, ComputeAtMap& ca_map)`, I think this checks for any inconsistency within the group on an unscheduled/untransformed fusion.

IIUC, this is different from what we are doing with `TransposeViewPropagator`, given that transpose scheduler doesn't require a global propagation across the fusion, but rather does it in three sections:
https://github.com/NVIDIA/Fuser/blob/e0f25481315450d4f9e8edf6a3f0cb5676d4a6ce/csrc/scheduler/transpose.cpp#L1004-L1008
https://github.com/NVIDIA/Fuser/blob/e0f25481315450d4f9e8edf6a3f0cb5676d4a6ce/csrc/scheduler/transpose.cpp#L1035-L1043
https://github.com/NVIDIA/Fuser/blob/e0f25481315450d4f9e8edf6a3f0cb5676d4a6ce/csrc/scheduler/transpose.cpp#L1123-L1133

> Is this check used as a quick alternative?

In short, yes.
I think having a `requiresForwardViewReplay` like analysis would be more precise, while our `TransposeViewPropagator` here is like a sledgehammer that probably rejects more fusion than what's necessary. But it's going to be harder to have an accurate analysis like `requiresForwardViewReplay` for transpose.

Reply:
I think it's fine as long as it erroneously fails to reject invalid cases. 

I wonder what happens if a fusion has a reshape after a transpose that is not compatible with the transpose schedule. For example:

```
tv0 = [1024, 4, 8]; // small inner domains
tv1 = transpose(tv0, 2, 3); // [1024, 8, 4]
tv2 = reshape(tv1, [1024, 8, 4], [1024, 8, 2, 2]); 
```

Reply:
That's a good one. We should reject fusion with incoherent transform introduced by reshape.

The analysis is hard, so we just don't support `small inner domain` + `reshape` at this point. https://github.com/NVIDIA/Fuser/blob/bebdf1b027e2cdae0cb0356b10111cdebaffb51a/csrc/scheduler/transpose.cpp#L667-L669

Reply:
Oh, I see, so at least with this restriction we should never do a merge of inner domains.

How about a case where we have large enough innermost domains and they are merged by a reshape. The transpose scheduler would want to split them, which seems to conflict with the reshape, no?

Reply:
My naive understanding is that, incoherent transform applies only when we try to *undo* a previous transformation. i.e. if we split a previously merged dimension with the same factor?!

The split on innermost dimension in transpose scheduler is going to be a constant factor for tiling. Which "seems" to be not causing any issue here.

```
Inputs:
  T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ], float
Outputs:
  T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ], float

%kernel_math {
T1_l[ iS5{i3}, iS4{i2}, iS3{i0} ]
   = Set.Permute( T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ] )
T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ] = view( T1_l[ iS5{i3}, iS4{i2}, iS3{i0} ] )
}

} // {Re-written complete fusion}
Segmented_Fusion Dump: -- fusion segments:
Segmented_Fusion{
groups:
g{0, 3}

edges:

group details:
g{(transpose)
inputs:
T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ] float
outputs:
T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ] float


T1_l[ iS5{i3}, iS4{i2}, iS3{i0} ]
   = Set.Permute( T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ] )
(0)
T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ] = view( T1_l[ iS5{i3}, iS4{i2}, iS3{i0} ] )
(3)
}

===== Transpose Stats ========
inputs: T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ]
outputs: T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ]
shape: 1024 8192
num_elems: 8388608
n_io_tensors: 2
max_io_dtype_size: 4
group 1: T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ]
reference1: T2_g[ iS6{i3}, iS11{( i2 * i0 )}rf ]
inner_most_id1 position: 1 (in reference 1)
group 2: T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ]
reference2: T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ]
inner_most_id2 position: 0 (in reference 1)


===== Transpose Parameters ========
Tag: Transpose heuristics Transpose Characteristics:
 BlckX: 128
 input tile size: 32
 output tile size: 32
 elements per tile: 1024
 elements per thread: 8
Vectorize group 1, Factor: 4
Unroll group 1, Factor: 2
Unroll group 2, Factor: 8
====================================
```

```
  auto tv0 = makeSymbolicTensor(3);
  fusion->addInput(tv0);

  auto tv1 = transpose(tv0, 0, 2);
  auto tv2 = reshape(tv1, {1024, 8, 1024}, {1024, 8192});
  fusion->addOutput(tv2);
```

Although, looks like vectorization isn't enabled on output.... something isn't right there. :cry:

Reply:
^^^ I'll open an issue tracking this one.
Do you also want me to open another issue to book keeping our concern regarding the potential incoherent transform for reshape + transpose scheduler tiling?

Reply:
OK, I thought the reference tensor for the post transpose part would be the transpose output, but that isn't the case. Here, `tv2` is the one, so there should be no issue. 

I think the check here is sufficient for now, but yes, please add some more comments with an issue to make sure we don't forget.